service/history/shard/context_impl.go

```diff
@@ -559,7 +559,11 @@ func (s *ContextImpl) CreateWorkflowExecution(
 
 	currentRangeID := s.getRangeIDLocked()
 	request.RangeID = currentRangeID
+
+	// Drop lock while calling persistence to allow for more concurrency
+	s.wUnlock()
 	resp, err := s.executionManager.CreateWorkflowExecution(request)
+	s.wLock()
```

Comment:
instead of locking here again, can we make 

`updateMaxReadLevelLocked` & `transitionLocked` lock free?

Reply:
e.g. using for loop & atomic load & atomic compare and swap

Reply:
I think that would make this code harder to reason about and maintain. I don't think we should reach for atomics unless we have good data that mutexes are a real performance issue in a particular spot

Reply:
partially agree with ^, but still, unlock & lock again is weird.

seems that create workflow / update workflow / conflict resolve workflow / add tasks all need locks due to
a. allocateTaskIDsLocked: allocation of task IDs
b. update shard's state and / or max readable task ID

`a` can be moved into [transaction manager](https://github.com/temporalio/temporal/blob/master/service/history/workflow/transaction_impl.go)
then the situation of lock & unlock & lock & unlock situation can be solved?

Reply:
there is actually one issue

the lock originally covers the entire task ID allocation & DB write & update max readable task ID

now, the lock is split into 2 sections, one section covers the task ID allocation, another section covers the update of max readable task ID
above split of lock will cause issue

e.g.
t = 0 max readable task ID is 0 (<= 0)
t = 0 thread 1 gets task ID 1-5
t = 1 thread 2 gets task ID 6-10
t = 2 thread 2 gets the lock to write to DB and successfully persisted the mutable state & tasks
t = 3 thread 2 update the max readable task ID to 10 (<= 10)

at this moment tasks with ID 1-5 is not persisted

t = 4 task reader & task processor sees task <= 10 are readable and load tasks
t = 5 all tasks <= 10 are loaded and processed, task reader will begin to read from 10 (>10)
t = 6 thread 1 gets the lock to write to DB and successfully persisted the mutable state & tasks

NOTICE: tasks with ID 1-5 will never be read from DB (see t = 5)

Reply:
there should be a utility 
* either making sure whoever get the allocated task IDs first also write to DB first
* or tracking allocated but not used task IDs and preventing updating of max readable task ID if `allocated but not used task IDs` exists